### DISCUSSION TITLE
Using association table to describe a Many-to-one relationship

### DISCUSSION THREAD

**Discussion by mattalbr** [sqlalchemy/sqlalchemy#8602] [Usage Questions] [answered]
So, we have a relationship that *at the moment* appears to be many-to-one, but we suspect might reveal itself to be many-to-many in a few months time. To simplify a potential DB migration if that turns out to be the case, we decided to create an association table with a unique constraint on one of the columns. The idea is if we end up converting to many-to-many, we update our code and then simply remove the unique constraint on the association table, without any data migrations. That code looks like

```python
class Child(base):
  __tablename__ = "child"
  id = Column(Integer, primary_key=True)
  parent = relationship("Parent", secondary="child_parent", back_populates="children", uselist=False)

class Parent(base):
  __tablename__ = "parent"
  id = Column(Integer, primary_key=True)
  children = relationship("Child", secondary="child_parent", back_populates="parent")

child_parent = sa.Table(
  "child_parent", database.metadata_obj,
  sa.Column("child_id", sa.ForeignKey('child.id'), primary_key=True, unique=True),
  sa.Column('parent_id', sa.ForeignKey('parent.id'), primary_key=True, index=True))
```

The problem is that it appears sqlalchemy coins this a `MANYTOMANY` relationship despite setting uselist=False. This throws off any inspection-based libraries, e.g. sqladmin. I'm unsure if this is a bug, an unsupported use case, or if there's some workaround for me to hint to sqlalchemy that this relationship is `MANYTOONE`, not `MANYTOMANY`

**Comment by zzzeek** [comment, marked as answer]
if you have "secondary", it's MANY TO MANY as far as the ORM is concerned, it just happens to be uselist=False which means it will only expect one entry.   im not sure what sqladmin is, but it should be looking at uselist also.

**Reply by mattalbr** [reply]
I guess I don't see why it's fundamentally different from the `ONETOONE` example in the docs where there's a foreign key without a unique constraint, so purely based on the DB arrangement it should be `ONETOMANY`, but uselist=False turns it into a `ONETOONE`.

I don't mind trying to upstream changes to any libraries we use that don't properly handle uselist, it just seems strange to me that sometimes uselist changes the relationship type (i.e. `ONETOONE`) and sometimes, as in this case, it doesn't.

**Reply by zzzeek** [reply]
sqlalchemy doesnt have a ONETOONE constant.  that is also something that is only determinable by looking at the uselist parameter.   uselist never changes that constant, it is determined strictly by the location of the "foreign key" columns:   if on the "local" side, its MANYTOONE.  if on the "remote" side, it's ONETOMANY.  if on a "secondary" table, it's MANYTOMANY.  Those are the three choices.  then uselist adds python-only behaviors of "use a collection" or "use a scalar".

so if you have a tool that is detecting "one to one", it's looking at uselist.  

for many-to-many plus uselist, I guess there is some ambiguity there if that's "one to one" or "many to one", but looking at just the one side, it's "many to one".

**Reply by mattalbr** [reply]
Got it, super helpful. Thank you!
